### PR TITLE
Audit: Use existing cloned object for hashing (don't re-clone)

### PR DESCRIPTION
**Pull request by peteski22**
### Description

As each formatter node in an audit pipeline will `Clone` the `LogInput` pointer (https://github.com/hashicorp/vault/blob/main/audit/entry_formatter.go#L123-L124) before making any mutations, we no longer need to deal with dereferencing pointers and using `copystructure` deeper down inside the hashing functions (`hashAuth`, `hashRequest`, `hashResponse`). 

We can now just modify the parts of `LogInput` as we want within `hashstructure.go` vs. copying and returning a new value (not the best code smell, but it makes the change easier to understand at the moment).

There are also `nil` checks so we can call them safely and nothing will be mutated if there's nothing to mutate.

Added some go docs to explain what's happening, but it's all contained within the `audit` package.

### HashiCorp Checklist

- [x] **Labels:** If this PR is the CE portion of an ENT change, and that ENT change is
  getting backported to N-2, use the new style `backport/ent/x.x.x+ent` labels
  instead of the old style `backport/x.x.x` labels.
- [x] **Labels:** If this PR is a CE only change, it can only be backported to N, so use
  the normal `backport/x.x.x` label (there should be only 1).
- [x] **ENT Breakage:** If this PR either 1) removes a public function OR 2) changes the signature
  of a public function, even if that change is in a CE file, _double check_ that
  applying the patch for this PR to the ENT repo and running tests doesn't
  break any tests. Sometimes ENT only tests rely on public functions in CE
  files.
- [x] **Jira:** If this change has an associated Jira, it's referenced either
  in the PR description, commit message, or branch name.
- [x] **RFC:** If this change has an associated RFC, please link it in the description.
- [x] **ENT PR:** If this change has an associated ENT PR, please link it in the
  description. Also, make sure the changelog is in this PR, _not_ in your ENT PR.

### CODE REVIEW

File: audit/hashstructure.go
Comment (raskchanky) [inline audit/hashstructure.go:31]: Godocs that explain all the arguments are _the best_.

File: audit/hashstructure_test.go
Comment (raskchanky) [inline audit/hashstructure_test.go:115]: You use `require.NoError(t, err)` for this in other tests. Maybe worth using here too?

**Review by raskchanky — APPROVED** [review]
I don't know the audit code very well, but I this followed what you're doing here and it seems fine. It might be worth waiting for a review from someone that knows this a bit better than me though.

**Comment by raskchanky** [comment]
CleanShot 2024-07-30 at 13 46 48

Also JUST LOOK AT THAT CI

File: audit/hashstructure_test.go
Comment (peteski22) [reply audit/hashstructure_test.go:115]: Yeah, I think I'm a bit stuck as this one is implementing the `Salter` interface's `Salt` method, so I can't adjust the func params to include `*testing.T`. Good shout though.

**Review by kubawi — APPROVED** [review]
LGTM

**Pull request merged**
